test/util/thread.test.cpp

```diff
@@ -328,3 +328,46 @@ TEST(Thread, DeleteBeforeChildStarts) {
     // Should process the queue before destruction.
     ASSERT_TRUE(flag);
 }
+
+TEST(Thread, PoolWait) {
+    auto pool = Scheduler::GetBackground();
+
+    constexpr int threadCount = 10;
+    for (int i = 0; i < threadCount; ++i) {
+        pool->schedule([&] { std::this_thread::sleep_for(std::chrono::milliseconds(100)); });
+    }
+
+    EXPECT_EQ(0, pool->waitForEmpty(std::chrono::milliseconds(500)));
+}
+
+TEST(Thread, PoolWaitTimeout) {
+    auto pool = Scheduler::GetBackground();
+
+    pool->schedule([&] { std::this_thread::sleep_for(std::chrono::milliseconds(100)); });
+
+    EXPECT_EQ(1, pool->waitForEmpty(std::chrono::milliseconds(10)));
```

Comment:
I'm wondering if we'd ever see the OS scheduler preempt this test in a way that could cause a failure.

Reply:
I made the task block on a mutex so there's no potential for order ambiguity.